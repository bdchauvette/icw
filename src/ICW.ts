import { DelegatingAsyncIterable } from "./__internal__/DelegatingAsyncIterable";
import { of } from "./of";
import { from, FromInput } from "./from";

export class ICW<T> extends DelegatingAsyncIterable<T> {
  static from<U>(input: FromInput<U>): ICW<U> {
    return new ICW(from(input));
  }

  static of<U>(...items: U[]): ICW<U> {
    return new ICW(of(...items));
  }
}
