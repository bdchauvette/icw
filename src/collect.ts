export async function* collect<T>(
  iterable: AsyncIterable<T> | Iterable<T>
): AsyncIterable<T[]> {
  let collectedResults = [];
  for await (let result of iterable) collectedResults.push(result);
  yield collectedResults;
}
