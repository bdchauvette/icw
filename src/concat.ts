import { flat } from "./flat";
import { toIterable } from "./__internal__/toIterable";
import { IterableLike } from "./IterableLike";

export async function* concat<T, U>(
  iterableLike: IterableLike<T>,
  ...values: U[]
): AsyncIterableIterator<T | U> {
  yield* toIterable(iterableLike);
  yield* flat(values, 1);
}
