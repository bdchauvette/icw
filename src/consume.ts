export const consume = async <T>(
  iterable: Iterable<T> | AsyncIterable<T>
): Promise<void> => {
  // @ts-ignore
  for await (let result of iterable) continue;
};
