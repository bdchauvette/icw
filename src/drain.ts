export async function drain<T>(
  iterable: AsyncIterable<T> | Iterable<T>
): Promise<void> {
  // @ts-ignore
  for await (let result of iterable) continue;
}
