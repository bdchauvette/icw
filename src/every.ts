import { IterableLike } from "./IterableLike";
import { withIndex } from "./withIndex";

export async function every<T>(
  iterableLike: IterableLike<T>,
  predicate: (value: T, index?: number) => boolean | Promise<boolean>,
  thisArg?: any
): Promise<boolean> {
  let iterableDidPass = false;

  for await (let [value, index] of withIndex(iterableLike)) {
    iterableDidPass = Reflect.apply(predicate, thisArg, [value, index]);
    if (!(await iterableDidPass)) return false;
  }

  return iterableDidPass;
}
