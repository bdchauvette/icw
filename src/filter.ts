import { withIndex } from "./withIndex";

/**
 * Yields results from the input iterable that satisfy the provided condition.
 */
export async function* filter<T>(
  iterable: AsyncIterable<T> | Iterable<T>,
  shouldInclude: (result: T, index?: number) => boolean | Promise<boolean>,
  thisArg?: any
): AsyncIterable<T> {
  for await (let [result, index] of withIndex(iterable)) {
    let resultIsOk = await shouldInclude.call(thisArg, result, index);
    if (resultIsOk) yield result;
  }
}
