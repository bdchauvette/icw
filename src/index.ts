export { ICW } from "./ICW";

export { from } from "./from";
export { of } from "./of";

export { drain } from "./drain";
export { filter } from "./filter";
export { map } from "./map";
export { reject } from "./reject";
export { scan } from "./scan";
export { withIndex } from "./withIndex";
