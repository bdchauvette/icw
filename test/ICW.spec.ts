import { ICW } from "../src/ICW";
import { runOfSuite } from "./suites/runOfSuite";
import { runFromSuite } from "./suites/runFromSuite";

test("is a class", () => {
  expect.assertions(2);
  expect(ICW).toBeFunction();
  expect(String(ICW)).toMatch(/^class/);
});

test("is an async iterable", async () => {
  expect.assertions(1);
  await expect(new ICW([])).toBeAsyncIterable();
});

describe("::from", () => {
  test("is a function", () => {
    expect.assertions(1);
    expect(ICW.from).toBeFunction();
  });

  test("returns an ICW instance", () => {
    expect.assertions(1);
    expect(ICW.from([])).toBeInstanceOf(ICW);
  });

  runFromSuite(ICW.from);
});

describe("::of", () => {
  test("is a function", () => {
    expect.assertions(1);
    expect(ICW.of).toBeFunction();
  });

  test("returns an ICW instance", () => {
    expect.assertions(1);
    expect(ICW.of()).toBeInstanceOf(ICW);
  });

  runOfSuite(ICW.of);
});
  });

  ofSuite(ICW.of);
});
});
