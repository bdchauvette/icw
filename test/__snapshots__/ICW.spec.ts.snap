// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`::from throws an error when input is "Array-like with negative index" 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;

exports[`::from throws an error when input is "Array-like with too large an index" 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;

exports[`::from throws an error when input is "Symbol" 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;

exports[`::from throws an error when input is "non-iterable object" 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;

exports[`::from throws an error when input is "number" 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;

exports[`::from throws an error when input is false 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;

exports[`::from throws an error when input is null 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;

exports[`::from throws an error when input is true 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;

exports[`::from throws an error when input is undefined 1`] = `"Must provide an iterable, an Array-like value, or a Promise."`;
